### PR TITLE
Android: Improve UI scaling on smaller high-density displays

### PR DESCRIPTION
**Pull request by stujones11**
In conjunction with #7784  this PR attempts to fix all the remaining menu scaling issues on mobile devices with smaller screens and higher display densities.

I found that most of the problem was with how `guiFormSpecMenu.cpp` determines the ideal scaling factor. For Android, I believe that it is always best to make any form as big as reasonably comfortable.

This PR also includes a fix for #2294  this may seem a bit overkill but I did not find a better way to upscale the checkbox tick texture. The images were exported from Inkscape using a SVG I created myself, which I  can also include if desired.

I have also taken the liberty of fixing some alignment issues with the touchscreen threshold on the main menu settings tab that was annoying me and has been pointed out elsewhere: https://github.com/minetest/minetest/pull/7784#issuecomment-429661081

Note that the following screenshots have been scaled down and composited to save bandwidth so disregard the image quality and focus on the actual scaling and checkboxes.

The build from which these were taken also had  #7820 applied, hence why you don't see my nav-bar. That does make a significant improvement for me, combined with these changes.

All screenshots with `gui_scaling = 1.0`
Device: Huawei honor 9 lite

screens

### CODE REVIEW

**Comment by juozaspo** [comment]
Done some tests in my Huawei Honor 7 Lite (Android 7.0). On xxhdpi screen with 1920x1080 resolution (480 dpi) forms appear ok but some fonts are still too large (not the scope of this PR). Build made with ndk r17 & latest android sdk available. Screenshots:
<details>

screenshot_20181103-081436
screenshot_20181103-081508
screenshot_20181103-081533
screenshot_20181103-081542
screenshot_20181103-081638
screenshot_20181103-081658

</details>

My screen info (copied from https://github.com/minetest/minetest/pull/7784#issuecomment-430131919):
```
Model: HUAWEI NEM-L21
Android version: 7.0 - N
API level: 24
Drawable density: xxhdpi
Layout size: normal
Screen (dpi): 480
Width (px): 1080
Height (px): 1920
Usable width (px): 1080
Usable height (px): 1812
Width (dp): 360
Height (dp): 640
Smallest dp: 360
Orientation: Portrait
Orientation degree: 0°
Diagonal Size(inches): 5.0
X dpi: 422,0
Y dpi: 424,1
Touchscreen: Finger
Refresh rate (fps): 60,0
```

**Comment by stujones11** [comment]
@juozaspo Thank you for testing, you should find that it looks better still if you also apply #7820. The nav-bar hiding makes a significant improvement on my honor 9 lite.

I am finding a font size of 14 about right for me with this, while previously 13 was a better fit. The good thing is that all the forms are now useable enough to alter this.

**Comment by stujones11** [comment]
Now includes a fix for #7798 and scaled the dropdown buttons relative to display density.

File: src/client/clientlauncher.cpp [hunk 1]
```diff
@@ -127,7 +130,30 @@ bool ClientLauncher::run(GameParams &game_params, const Settings &cmd_args)
...
-
+#ifdef __ANDROID__
+	float density = porting::getDisplayDensity();
+	skin->setSize(gui::EGDS_CHECK_BOX_WIDTH, (s32)(17.0f * density));
+	skin->setSize(gui::EGDS_SCROLLBAR_SIZE, (s32)(15.0f * density));
+	skin->setSize(gui::EGDS_WINDOW_BUTTON_WIDTH, (s32)(15.0f * density));
+	if (density > 1.5f) {
+		std::string sprite_path = porting::path_user + "/textures/base/pack/";
+		if (density > 3.5f)
+			sprite_path += "checkbox_64.png";
```
Comment (nerzhul) [inline src/client/clientlauncher.cpp]: use append instead of += operator for performance issues

File: src/client/clientlauncher.cpp [hunk 2]
```diff
@@ -127,7 +130,30 @@ bool ClientLauncher::run(GameParams &game_params, const Settings &cmd_args)
...
+		else if (density > 2.0f)
+			sprite_path += "checkbox_32.png";
+		else
+			sprite_path += "checkbox_16.png";
+		// Texture dimensions should be a power of 2
+		gui::IGUISpriteBank *sprites = skin->getSpriteBank();
+		video::IVideoDriver *driver = RenderingEngine::get_video_driver();
+		video::ITexture *sprite_texture = driver->getTexture(sprite_path.c_str());
```
Comment (nerzhul) [inline src/client/clientlauncher.cpp]: you can inline that in the conditional directly

File: src/client/clientlauncher.cpp (hunk 1 above)
Comment (stujones11) [reply src/client/clientlauncher.cpp]: Done, I agree it is better but there really is no performance issue here. This is the client launcher and (afaik) it runs only once per client session ;)

File: src/client/clientlauncher.cpp (hunk 2 above)
Comment (stujones11) [reply src/client/clientlauncher.cpp]: > you can inline that in the conditional directly

Sorry, I have no idea what you mean by this. Please show me how you would like it.

**Comment by stujones11** [comment]
@juozaspo With this PR, all scrollbar should now be a fixed size, based on display density. However, this does not really make the content store any more readable. Fixing that is well beyond the scope of this PR but I believe there are issues/PR for this. It's just as much an issue with the default desktop window.

File: src/client/clientlauncher.cpp (hunk 2 above)
Comment (SmallJoker) [reply src/client/clientlauncher.cpp]: I think he meant it like this:
```C++
if (video::ITexture *sprite_texture = driver->getTexture(sprite_path.c_str())) {
	// Use sprite_texture
```
This doesn't look great either, so you may just keep it as-is.

Comment (stujones11) [reply src/client/clientlauncher.cpp]: Ah, I see he means now but surely it looks better like this.

**Comment by rubenwardy** [comment]
The descriptions in the store were never intended to be scrollable. Ideally, there should be a label element with support for wrapping and truncating text with `…`

**Comment by juozaspo** [comment]
Made a new test build made from a52165dbd3b887d373b29656cc9e03534329f235; There's some new screenshots for things that got changed since my previous test. Notice how serverlist scrollbar got a bit smaller:
<details>






</details>

**Comment by stujones11** [comment]
> Notice how serverlist scrollbar got a bit narrower

Previously the `guiTable` would scale with the form so I have made a slight change that _should_ restore the original behaviour. This also allows me to reduce the default size for the automatic one you see on drop-downs and text overflows. It looks about right on both of my test devices, one `ldpi` and the other `xxhdpi` like your own.

Thanks again for testing.

**Comment by juozaspo** [comment]
On my last test build (ce6d559f4cdc4571ffb2cd85ac64adc6989fcced) the server list scrollbar appears like in my first test, so screenshot not needed to post; Though I'd prefer scrollbar buttons having same size everywhere (consistant) but this version looks good too :)

**Comment by rubenwardy** [comment]
My laptop is broken so I can't check - what does a `size[1,1]` formspec look like?

**Comment by stujones11** [comment]
> what does a size[1,1] formspec look like?

With this PR, exactly the same as a size[10,10] formspec ;)

screen

File: src/gui/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -284,12 +284,13 @@ void GUIFormSpecMenu::parseSize(parserData* data, const std::string &element)
 		data->invsize.Y = MYMAX(0, stof(parts[1]));
 
 		lockSize(false);
+#ifndef __ANDROID__
 		if (parts.size() == 3) {
 			if (parts[2] == "true") {
 				lockSize(true,v2u32(800,600));
```
Comment (paramat) [inline src/gui/guiFormSpecMenu.cpp]: 800x600 was our previous default desktop screensize, does this need updating to the new default of 1024x600?

File: src/gui/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -2145,6 +2147,16 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
...
 			// multiplied by gui_scaling, even if this means
 			// the form doesn't fit the screen.
+#ifdef __ANDROID__
+			// For mobile devices these magic numbers are
+			// different and forms should always use the
+			// maximum screen space available.
+			double fitx_imgsize = mydata.screensize.X /
+				((12.0/8.0) * (0.5 + mydata.invsize.X));
+			double fity_imgsize = mydata.screensize.Y /
+				((15.0/11.0) * (0.85 + mydata.invsize.Y));
```
Comment (paramat) [inline src/gui/guiFormSpecMenu.cpp]: Here and above, spaces around '/'

File: src/gui/guiTable.cpp [hunk 5]
```diff
@@ -78,10 +78,12 @@ GUITable::GUITable(gui::IGUIEnvironment *env,
...
 	updateAbsolutePosition();
-
+	float density = RenderingEngine::getDisplayDensity();
+#ifdef __ANDROID__
+	density = 1; // dp scaling is applied by the skin
+#endif
 	core::rect<s32> relative_rect = m_scrollbar->getRelativePosition();
-	s32 width = (relative_rect.getWidth()/(2.0/3.0)) *
-			RenderingEngine::getDisplayDensity() *
+	s32 width = (relative_rect.getWidth()/(2.0/3.0)) * density *
```
Comment (paramat) [inline src/gui/guiTable.cpp]: Spaces around '/'

File: src/gui/guiFormSpecMenu.cpp (hunk 3 above)
Comment (paramat) [reply src/gui/guiFormSpecMenu.cpp]: If changed, spaces after the commas if possible.

File: src/gui/guiFormSpecMenu.cpp [hunk 6]
```diff
@@ -437,10 +438,11 @@ void GUIFormSpecMenu::parseCheckbox(parserData* data, const std::string &element
 			fselected = true;
 
 		std::wstring wlabel = translate_string(utf8_to_wide(unescape_string(label)));
+		s32 spacing = Environment->getSkin()->getSize(gui::EGDS_CHECK_BOX_WIDTH) + 7;
 
 		core::rect<s32> rect = core::rect<s32>(
 				pos.X, pos.Y + ((imgsize.Y/2) - m_btn_height),
-				pos.X + m_font->getDimension(wlabel.c_str()).Width + 25, // text size + size of checkbox
+				pos.X + m_font->getDimension(wlabel.c_str()).Width + spacing,
 				pos.Y + ((imgsize.Y/2) + m_btn_height));
```
Comment (paramat) [inline src/gui/guiFormSpecMenu.cpp]: Since you're altering this code block, spaces around '/' here and above.

**Comment by paramat** [comment]
> For Android, I believe that it is always best to make any form as big as reasonably comfortable.

Agreed.

File: src/gui/guiFormSpecMenu.cpp (hunk 3 above)
Comment (stujones11) [reply src/gui/guiFormSpecMenu.cpp]: > does this need updating to the new default of 1024x600?

I guess it probably should be for desktop but I am not sure changing that is this PR is a good idea. I am not even entirely sure of its purpose and there may be a risk of it effecting the look of some menus/forms. As you can see, this section is simply ignored for android.

**Comment by paramat** [comment]
@stujones11 do you have an apk available which contains just this PR, for people to test?

**Comment by stujones11** [comment]
Here is a dubug apk built from fde1749e47344e00f974e9a977879cdf3bcaf213 however, you should keep in mind that this is improved further by #7820 & #7850. Also note that the current default `gui_sacling` is 1.1 whereas the PR is based on 1.0 which should be introduced #7784

**Download:** https://www.dropbox.com/s/kgtwew5jlrw4cw1/Minetest-debug.apk?dl=0

If you have an existing installation of minetest on your device then it is important that you rename your current `Minetest` directory before installing this.

**Comment by rubenwardy** [comment]
> @rubenwardy With this PR, exactly the same as a size[10,10] formspec ;)

I think it's probably better to have a max size rather than making every formspec the same size

**Comment by stujones11** [comment]
> I think it's probably better to have a max size rather than making every formspec the same size

This would be quite difficult to implement and perfect across such a wide range of sizes and densities, largely because of how formspecs are scaled. In reality, there are no 1x1 forms and I see no good reason to make forms any smaller, particularly on mobile phones.

This is simple solution to a fairly complex problem that does not require any factoring for density which should therefore work consistently well on any mobile device.

**Comment by stujones11** [comment]
After giving this some more thought, I do agree that a single button form should not need fill the entire screen. I have now made the preferred inventory image size 1/10 the screen height.

Here is how a 5x1 form looks, simulated on desktop at 1080x540px

**Before:**

screenshot_20181114_193751

**After:**

screenshot_20181114_192139

**Comment by stujones11** [comment]
@IhrFussel Thank you for your feedback. The default `gui_scaling` should remain at 1.0 following #7784

**Comment by paramat** [comment]
Seems like enough testing, i'll check the code and probably +1 soon.

**Comment by paramat** [comment]
Although i don't have a complete understanding of the code, it looks ok as far as i can tell, and i think the testing counts for something, so :+1:

**Comment by nerzhul** [comment]
Thanks for your efforts

**Comment by stujones11** [comment]
> However, with unified_inventory there's still a free 5mm border

This border size will depend mostly on the aspect fit of the from to that of the display so we need to allow a little bit space for this, I think. Also consider those using cases/protectors.

Good to see this merged, thanks guys!

**Pull request merged**